client/src/components/annotators/ImageAnnotator.vue

```diff
@@ -1,258 +1,290 @@
+      const min = Math.floor(Math.max(0, data.frame - cachedFrames * (1 - local.frontBackRatio)));
+      const max = Math.ceil(
+        Math.min(data.frame + local.frontBackRatio * cachedFrames, data.maxFrame),
+      );
+      const frameDiff = Math.abs(data.frame - local.lastFrame);
+      const prevFrame = (data.frame < local.lastFrame);
+      local.pendingImgs.forEach((imgInternal) => {
         // the current loading cache needs to be wiped out if we seek forward, backwards or
         // if we are out of the current range of the cache
-        if (imageAndFrame[1] < min || imageAndFrame[1] > max || frameDiff > 1 || prevFrame) {
-          imgs[imageAndFrame[1]] = null;
+        if (imgInternal.frame < min || imgInternal.frame > max || frameDiff > 1 || prevFrame) {
+          // Removal from list indicates "we are no longer attempting to load this image"
+          local.imgs[imgInternal.frame] = undefined;
+          // unset src to cancel outstanding load request
           // eslint-disable-next-line no-param-reassign
-          imageAndFrame[0].src = '';
-          this.pendingImgs.delete(imageAndFrame);
+          imgInternal.image.src = '';
+          local.pendingImgs.delete(imgInternal);
         }
       });
       // if not playing we want the seeked to frame immediately and prevent caching until loaded
-      if (!this.playing && !imgs[frame] && frame > 0) {
-        await this.loadFrame(frame);
+      if (!data.playing && !local.imgs[data.frame] && data.frame > 0) {
+        await cacheFrame(data.frame)?.onloadPromise;
       }
       // Cache a new range of images based on current frame
-      return this.cacheNewRange(min, max);
-    },
-    /**
-     * Wraps loading of a single frame in a promise, used to gurantee frame loads
-     * before execution of caching.
-     * @param {int} frame number to be loaded
-     * @returns {Promise} resolves when the image from the frame is loaded
-     */
-    loadFrame(frame) {
-      return new Promise((resolve) => {
-        const img = new Image();
-        img.crossOrigin = 'Anonymous';
-        this.imgs[frame] = img;
-        img.onload = () => {
-          img.onload = null;
-          img.cached = true;
-          img.frame = frame;
-          resolve(frame);
-        };
-        this.loadImageFunc(this.imageData[frame], img);
-      });
-    },
-    /**
-     * Caches a new range of frames to load in a forward->back pattern from the current frame
-     * This allows for easily seeking backwards after seeking initially
-     * @param {int} min lower bound frame number for caching
-     * @param {int} max upper bound frame number for caching
-     */
-    cacheNewRange(min, max) {
-      const frameImages = [];
-      for (let i = this.frame; i <= max; i += 1) {
-        frameImages[i] = this.cacheFrame(i);
-        const minusFrame = this.frame - (i - this.frame);
-        if (minusFrame >= min) {
-          this.cacheFrame(minusFrame);
+      cacheNewRange(min, max);
```

Comment:
Curiosity here. It looks like cancelling the single frame that is seeked to before loaded allows it to fall through and cacheNewRange.  cacheNewRange should only be called on on the current frame if it is loaded?  Testing makes it seem like it is called even if we jump before the frame is loaded.  May need a reset toggle on the function which will only cache the new range if the seeked frame loads first.